### ISSUE TITLE
Are events invoked in other threads?

### ISSUE THREAD

**Issue by iSLC** [brainboxdotcc/DPP#63]
If I bind a callback to an event (*`on_message_update` for example*). Is the callback guaranteed to be called in the same thread that created the cluster? I guess it doesn't since you start the cluster (*`true` argument*) and it has no way of getting back to the main thread. There's no control over the internal loop. Or way of integrating into parent loop (*like having an `update()` method called periodically*). I feel like this should be made a bit more obvious in case someone uses the library and deals with code that is not thread safe.

Because if I want to process that event in the main thread with a queue or something like that. I will receive a `message_update_t` argument in the event. But the problem now becomes: Who manages the lifetime of that `message_update_t` instance? So now I'm confronted with a few questions:

Do I block the thread that invoked the event and wait until it is handled in main thread? Seems weird.

Do I just queue the argument value and return? Then process it in main thread. But then who owns the `message* updated;` pointer? How do I tell the library that I'm still using that?

There are a bunch of raw pointers passed throughout the library and no way to get involved in their lifetime. No reference counting or something like that. And for something that deals with multiple threads seems weird.

Do I create deep copies of everything? Like keeping my own separate kind of cache?

I noticed in another issue that you said *`most discord bots are dedicated processes and not embedded in something else`*. That may be the case for some but not for most.

How would you approach this situation? How would you synchronize events and deal with the lifetime of objects that the library yields?

Thank you.

**EDIT**:

Upon further reading, I noticed that some of the objects that I was interested in their lifetime inherit from `dpp::managed` class.

Which are then down-casted to a `dpp::managed` pointer and stored inside various (_static/global_) containers in the `cache.cpp` unit. And then deleted by `garbage_collection()` which I guess is then called periodically from some other thread.

But in the `garbage_collection()` function I notice this line `delete g->first;` in which `g->first` is the `dpp::managed` pointer that was down-casted from whatever inherited that class. Which seems fine until you notice that `dpp::managed` does not have a virtual destructor.

Basically, `delete` does not know about the derived class destructor and only invokes the destructor of the `dpp::managed` class (_which has nothing to cleanup_). Doesn't this classify as a memory leak? ([link](https://stackoverflow.com/questions/7403883/derived-class-with-non-virtual-destructor))

Regarding the initial question. I guess I could modify the `dpp::managed` class to include a reference counter and the `garbage_collection()` function to take that into consideration before trying to delete the instance. Requirements to maintain this modification would be minimal since is just 2-3 lines of code.

EDIT2: But then `dpp::message` doesn't inherit from `dpp::managed` class. So I'm back to where I started. I guess I have to look further into this.

**Comment by braindigitalis** [comment]
dpp::managed shouldnt reference count -- i'll read this properly at some point when im not so busy and come back to you. Theres so much to read here!

**Comment by braindigitalis** [comment]
in short - dont try and hold onto any pointers youre given outside of the events.
Either in a lambda for one of the REST calls, or a websocket event. These are transient and will not last beyond the event. You should make copies of anything you want to store outside of the event. You shouldnt try to block or "control" any of the DPP threads.

There is a document that documents the thread model in the docs here: https://dpp.brainbox.cc/thread-model.html

The threading is relatively simple. there is no complex thread pool or similar. Every shard spawns an independent thread, that should only touch the resources of that shard. 

When you make a REST request (e.g. `dpp::cluster::guild_command_edit()`) the request is sent to a queue, which is a separate thread. This thread dispatches requests in-order, and when they complete, sends a signal to another thread which calls the completion lambdas for each event. This can mean that a very long running lambda can cause delays -- so don't do that!

**Comment by braindigitalis** [comment]
nice spot on the missing virtual dtor for dpp::managed, i'll fix this one now :)

**Comment by iSLC** [comment]
I've decided to go with the deep copy alternative for now. It shouldn't affect efficiency in any way atm since the thing that I'm using the library is something simple.

Thanks for the input.
